tests/services/test_cluster.py

```diff
@@ -0,0 +1,72 @@
+import itertools
+import pytest
+from unittest.mock import Mock, patch, call, AsyncMock
+from yapapi.services import Cluster, Service, ServiceError
+
+
+class _TestService(Service):
+    pass
+
+
+def _get_cluster():
+    return Cluster(engine=Mock(), service_class=_TestService, payload=Mock())
+
+
+@pytest.mark.parametrize(
+    "kwargs, calls, error",
+    [
+        (
+            {"num_instances": 1},
+            [call({})],
+            None,
+        ),
+        (
+            {"num_instances": 3},
+            [call({}) for _ in range(3)],
+            None,
+        ),
+        (
+            {"instance_params": [{}]},
+            [call({})],
+            None,
+        ),
+        (
+            {"instance_params": [{"n": 1}, {"n": 2}]},
+            [call({"n": 1}), call({"n": 2})],
+            None,
+        ),
+        (
+            # num_instances takes precedence
+            {"num_instances": 2, "instance_params": [{} for _ in range(3)]},
+            [call({}), call({})],
+            None,
+        ),
+        (
+            # num_instances takes precedence
+            {"num_instances": 3, "instance_params": ({"n": i} for i in itertools.count(1))},
+            [call({"n": 1}), call({"n": 2}), call({"n": 3})],
+            None,
+        ),
+        (
+            # num_instances takes precedence
+            {"num_instances": 4, "instance_params": [{} for _ in range(3)]},
+            [call({}) for _ in range(3)],
+            "`instance_params` iterable depleted after 3 spawned instances.",
+        ),
+    ],
+)
```

Comment:
I'd also include a test case for `num_instances = 0`

Reply:
added